### PR TITLE
Variables: Support static keys in AdHocFiltersVariable

### PR DESCRIPTION
**Pull request by ashharrison90**
**This PR requires a dashboard to be run in the Scenes architecture. Append `&scenes` query param to the dashboard URL in order to render it as a scene and see the static keys applied to adhoc filters.**

<details>
<summary>`dashboard.json` containing adhoc filter with statically defined keys for gdev-prometheus data source</summary>

```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 7375,
  "links": [],
  "panels": [
    {
      "datasource": {
        "type": "prometheus",
        "uid": "gdev-prometheus"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisBorderShow": false,
            "axisCenteredZero": false,
            "axisColorMode": "text",
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "insertNulls": false,
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 1,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom",
          "showLegend": true
        },
        "tooltip": {
          "maxHeight": 600,
          "mode": "single",
          "sort": "none"
        }
      },
      "pluginVersion": "11.0.0-pre",
      "targets": [
        {
          "datasource": {
            "type": "prometheus",
            "uid": "gdev-prometheus"
          },
          "editorMode": "code",
          "expr": "go_gc_duration_seconds{}",
          "instant": false,
          "legendFormat": "__auto",
          "range": true,
          "refId": "A"
        }
      ],
      "title": "Panel Title",
      "type": "timeseries"
    }
  ],
  "schemaVersion": 39,
  "tags": [],
  "templating": {
    "list": [
      {
        "datasource": {
          "type": "prometheus",
          "uid": "gdev-prometheus"
        },
        "defaultKeys": [
          {
            "text": "One",
            "value": "a"
          },
          {
            "text": "Two",
            "value": "b"
          }
        ],
        "filters": [],
        "name": "query0",
        "type": "adhoc"
      }
    ]
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "browser",
  "title": "static adhoc filters",
  "uid": "fdfmkc9aakphcf",
  "version": 2,
  "weekStart": ""
}
```
</details>

## How to use the new API to show adhoc filter with static dimensions in the provisioned dashboard?

Providing the statically defined list of defaultKeys via dashboard.json:

```ts
// dashboard.json 
{
  // ...
  "templating": {
    "list": [
      {
        "datasource": {
          "type": "prometheus",
          "uid": "gdev-prometheus"
        },
        "defaultKeys": [
          {
            "text": "One",
            "value": "a"
          },
          {
            "text": "Two",
            "value": "b"
          }
        ],
        "filters": [],
        "name": "query0",
        "type": "adhoc"
      }
    ]
  },
  // ...
}
```

## How to use the selected adhoc filters in the `DataSource.query` method? 

Queries targeting the same data source as the one configured for adhoc filter variable will receive the selected filters as  part of the `DataQueryRequest`, under `filters` property:

```ts
export interface DataQueryRequest<TQuery extends DataQuery = DataQuery> {
  requestId: string; // Used to identify results and optionally cancel the request in backendSrv

  ...  
  /** Filters to dynamically apply to all queries */
  filters?: AdHocVariableFilter[];

  ...
}
```

## Configuring the adhoc filter static keys from the UI.

From the UI point of view the MVP is just a code editor that allows you to enter a CSV content that will be mapped to dimension label and dimension id: 

image

**What is this feature?**

- adds support for providing static keys to `AdHocFiltersVariable`
- adds a basic switch and csv editor to the `AdHocVariableForm`
- fixes description in `GroupByVariableForm`

**Why do we need this feature?**

- to support static key definitions in `AdHocFiltersVariable`

**Who is this feature for?**

- anyone using `AdHocFiltersVariable`

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

- needs https://github.com/grafana/scenes/pull/612 releasing and then pulling into this PR 

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/dashboard-scene/serialization/transformSaveModelToScene.ts [hunk 1]
```diff
@@ -298,7 +298,12 @@ export function createSceneVariableFromVariableModel(variable: TypedVariableMode
       applyMode: 'auto',
       filters: variable.filters ?? [],
       baseFilters: variable.baseFilters ?? [],
-    });
+    };
+
+    if (variable.staticKeys) {
+      args.staticKeys = variable.staticKeys;
+    }
```
[the remark this answers is not in the document]
Comment (ashharrison90) [reply public/app/features/dashboard-scene/serialization/transformSaveModelToScene.ts:305]: ah yeah you're right, it's a hangover from a previous implementation i was trying 🤦 thanks!

**Comment by dprokop** [comment]
@ashharrison90 I think the key name/id are mesed up a bit. With the following keys setup:

```
one, a
two, b
```

the UI behaves like below, names are used as select values, but ids are used as the presented **selected** value:


https://github.com/grafana/grafana/assets/2376619/b1420627-b6c9-4323-9683-8aceb149fffe

**Review by dprokop — COMMENTED** [review]
@ashharrison90 - think this looks great. Other than the issue with dimension name/id in the picker, I think this is close to be mergable.

**Comment by ashharrison90** [comment]
> @ashharrison90 I think the key name/id are mesed up a bit. With the following keys setup:
> 
> ```
> one, a
> two, b
> ```
> 
> the UI behaves like below, names are used as select values, but ids are used as the presented **selected** value:
> 
>  Screen.Recording.2024-03-14.at.12.50.30.mov

addressed this with https://github.com/grafana/scenes/pull/645

**Comment by bergquist** [comment]
The input format might not be optimal. I don't want to exclude the possibility of having `,` as part of the label name/value. As long as the json definition stores the value separated it should be good enough for now.

**Review by dprokop — APPROVED** [review]

**Pull request merged**